src/frequenz/sdk/timeseries/battery_pool/_metric_calculator.py

```diff
@@ -0,0 +1,478 @@
+import logging
+from abc import ABC, abstractmethod
+from datetime import datetime, timezone
+from typing import Dict, Generic, Iterable, List, Optional, Set, TypeVar
+
+from ... import microgrid
+from ...microgrid.component import ComponentCategory, ComponentMetricId
+from ...timeseries import Sample
+from ._component_metrics import ComponentMetricsData
+from ._result_types import Bound, CapacityMetrics, PowerBoundsMetrics, SoCMetrics
+
+_logger = logging.getLogger(__name__)
+_MIN_TIMESTAMP = datetime.min.replace(tzinfo=timezone.utc)
+
+
+def battery_inverter_mapping(batteries: Iterable[int]) -> Dict[int, int]:
+    """Create mapping between battery and adjacent inverter.
+
+    Args:
+        batteries: Set of batteries
+
+    Returns:
+        Mapping between battery and adjacent inverter.
+    """
+    graph = microgrid.get().component_graph
+    bat_inv_map: Dict[int, int] = {}
+    for battery_id in batteries:
+        try:
+            predecessors = graph.predecessors(battery_id)
+        except KeyError as err:
+            _logger.error(str(err))
+            continue
+
+        inverter_id = next(
+            (
+                comp.component_id
+                for comp in predecessors
+                if comp.category == ComponentCategory.INVERTER
+            ),
+            None,
+        )
+        if inverter_id is None:
+            _logger.warning("Battery %d has no adjacent inverter.", battery_id)
+        else:
+            bat_inv_map[battery_id] = inverter_id
+    return bat_inv_map
+
+
+# Formula output types class have no common interface
+# Print all possible types here.
+T = TypeVar("T", Sample, SoCMetrics, CapacityMetrics, PowerBoundsMetrics)
+
+
+class MetricCalculator(ABC, Generic[T]):
+    """Define how to calculate high level metrics from many components data.
+
+    It specifies:
+        * what components and metrics its needs to calculate the results,
+        * how to calculate the result,
+    """
+
+    def __init__(self, batteries: Set[int]) -> None:
+        """Create class instance.
+
+        Args:
+            batteries: From what batteries the data should be aggregated.
+        """
+        self._batteries = batteries
+
+    @classmethod
+    @abstractmethod
+    def name(cls) -> str:
+        """Return name of the formula.
+
+        Returns:
+            Name of the formula
+        """
+
+    @property
+    def batteries(self) -> Set[int]:
+        """Return set of batteries that should be used to calculate the metrics.
+
+        Some batteries given in constructor can be discarded
+        because of the failing preconditions. This method returns set of
+        batteries that can be used in the calculator.
+
+        Returns:
+            Set of batteries that should be used.
+        """
+        return self._batteries
+
+    @property
+    @abstractmethod
+    def battery_metrics(self) -> Dict[int, List[ComponentMetricId]]:
```

Comment:
Maybe for this too, not sure how is it intended to be used, but maybe [`collections.abc.Mapping`](https://docs.python.org/3.11/library/collections.abc.html#collections.abc.Mapping) and [`collections.abc.Sequence`](https://docs.python.org/3.11/library/collections.abc.html#collections.abc.Sequence).

Etc...

Reply:
Done